File: src/snake_game_client/snake.py
from typing import List
import pygame

from game_config import GameConfig

class Snake:
    def __init__(self, x: int, y: int):
       self.head = (x, y)
       self.tail = []
       self.moves_history = []


    def get_move(self, direction: str):
       if direction == "U":
          return (-1, 0)
       elif direction == "D":
          return (1, 0)
       elif direction == "L":
          return (0, -1)
       elif direction == "R":
          return (0, 1)
       else:
          return (0, 0)
       
    def move(self, direction: str):
       move = self.get_move(direction)
       head_cpy = self.head
       self.head = (self.head[0] + move[0], self.head[1] + move[1])
       
       if len(self.tail) > 0:
         self.tail.pop()
         self.tail.insert(0, head_cpy)

    def get_last_snake_segment(self):
         if len(self.tail) > 0:
          return self.tail[-1]
         else:
          return self.head


    

File: python_cpp_binding/snake_game/scripts/Snake.cpp
#include "Snake.h"




Snake::Snake(int x, int y) {
    head = make_pair(x, y);
}

pair<int, int> Snake::get_move(char move) {
    switch (move) {
        case 'U':
            return make_pair(-1, 0);
        case 'D':
            return make_pair(1, 0);
        case 'L':
            return make_pair(0, -1);
        case 'R':
            return make_pair(0, 1);
        default:
            return make_pair(0, 0);
    }
}


void Snake::move_snake(char direction){
    pair<int, int> move_pair = get_move(direction);
    pair<int, int> head_cpy = head;

    head.first += move_pair.first;
    head.second += move_pair.second;

    if (tail.size() > 0){
        // get rid of the last segment
        tail.pop_back();
        // at the beginning of the tail insert the head
        tail.insert(tail.begin(), head_cpy);
    }
}

pair<int, int> Snake::get_last_snake_segment(){
    if (tail.size() == 0){
        return make_pair(head.first, head.second);
    }
    return make_pair(tail.back().first, tail.back().second);
}

char Snake::get_last_direction() {
    if (moves_history.size() < 2) {
        return ' '; 
    }
    pair<int, int> last_move = moves_history[moves_history.size() - 1];
    pair<int, int> second_last_move = moves_history[moves_history.size() - 2];

    if (last_move.first < second_last_move.first) {
        return 'U';
    } else if (last_move.first > second_last_move.first) {
        return 'D';
    } else if (last_move.second < second_last_move.second) {
        return 'L';
    } else {
        return 'R';
    }
}



